File: public/dsa/arrays/identical_twins/hashing_approach_2/identical_twins.cpp

#include <bits/stdc++.h>
using namespace std;

int getIdenticalTwinsCount(vector<int> &arr) {
    unordered_map<int, int> frequency;
    for (auto x : arr) {
        frequency[x]++;
    }
    int count = 0;
    for (auto x : frequency) {
        count += (x.second * (x.second - 1) / 2);
    }
    return count;
}
// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 2, 1};
    int count = getIdenticalTwinsCount(arr);
    cout << count << endl;
    return 0;
}
// tests-end

/*output
2
*/

File: public/dsa/arrays/identical_twins/hashing_approach_2/identical_twins.py
from collections import defaultdict
from typing import List


class Solution:
    def getIdenticalTwinsCount(self, arr: List[int]) -> int:
        mp = defaultdict(int)
        for v in arr:
            mp[v] += 1

        count = 0
        for k in mp:
            count += (mp[k] * (mp[k] - 1))//2

        return count


# tests-start
arr = [1, 2, 3, 2, 1]
sol = Solution()
print(sol.getIdenticalTwinsCount(arr))
# tests-end

'''output
2
'''
